### PR TITLE
Implement actual auto-approval of versions in auto_approve command

### LINKED ISSUE #4533: Auto-approve a fraction of WebExtension updates
(Edited to keep up with proposed changes)

We're currently experiencing significant delays in reviews due to decreased reviewer activity and increase submissions. It looks like this will only get worse leading to the transition to WebExtensions, so we need to start moving to a place where we automatically review at least some add-ons.

This is a first step that should hopefully be easy to implement quickly. For more context, please refer to [this draft PRD](https://docs.google.com/document/d/1VxvvzQQvfDOVCylSVlpfeXv0tDYRLZfIMvWOoraANGU/edit#).

Add-on submissions will be auto-approved based on a submission counter. In order to qualify for auto-approval, the add-ons need to meet this criteria:

1. Have been manually approved on initial submission.
2. Are not flagged for admin review. This can happen automatically on submission in some cases, but is generally done manually by reviewers.
3. WebExtensions (excludes Embedded WebExtensions).
4. Active Daily Users under a threshold. Initially 10K, but configurable.
5. Don’t trigger any of the following linter warnings:
5.1. CSP
5.2. Native Messaging
5.3. Content scripts for all URLs (not implemented yet).

1 out of every N consecutive updates for an add-on will be auto-approved, so a counter should be kept per-add-on. The counter shouldn’t change if a version is deleted by the developer, or if a new version is uploaded while the previous one was awaiting review (to avoid gaming via re-submission). The value N should be configurable, initially set to 5 and decreased over time if the outcome is positive. It can be set to 0 in order to shut down auto-approvals entirely.

Based on past performance, the initial setting should lead to ~40 auto-approved add-ons every week. It can be set to 0 in order to shut down auto-approvals.

The submission flow shouldn’t change. The last step should still tell the developer the add-on is pending review. Just like with unlisted add-ons, the review is performed automatically by the `Mozilla` user and a canned review message should be used both for the log and to send with the review email: `Thank you for your contribution.`. This is the same canned response used by reviewers when approving an add-on.

To ensure there aren’t serious security problems in the auto-approval system, the first implementation to go live should only identify the files that meet the auto-approval criteria and store it in the database, without auto-approving any files.

A report of identified files will be run manually so they can be verified by members of the security and review teams. Once the criteria is determined to be sufficient, auto-approvals can be turned on.

### PR DESCRIPTION
**Pull request by diox**
Fix #4533

### CODE REVIEW

File: src/olympia/editors/management/commands/auto_approve.py [hunk 1]
```diff
@@ -59,57 +66,77 @@ def handle(self, *args, **options):
...
+        # Is the addon already locked by a reviewer ?
+        if get_reviewing_cache(version.addon.pk):
+            self.stats['locked'] += 1
+            return
+
+        # If admin review or more information was requested, skip this
+        # version, let a human handle it.
+        if version.addon.admin_review or version.has_info_request:
+            self.stats['flagged'] += 1
+            return
```
Comment (EnTeQuAk) [inline src/olympia/editors/management/commands/auto_approve.py]: I wonder if we really should skip here and avoid gathering more stats about the approval process?

File: src/olympia/editors/management/commands/auto_approve.py [hunk 2]
```diff
@@ -59,57 +66,77 @@ def handle(self, *args, **options):
...
+            clear_reviewing_cache(version.addon.pk)
+
+    def approve(self, version):
+        """Do the approval itself, caling ReviewHelper to change the status,
+        sign the files, send the e-mail, etc."""
+        # Note: this should automatically use the TASK_USER_ID user.
+        helper = ReviewHelper(addon=version.addon, version=version)
+        helper.handler.data = {
+            'comments': _(u'This version has been approved for the public.'
+                          u'\r\n\r\nThank you!')
```
Comment (EnTeQuAk) [inline src/olympia/editors/management/commands/auto_approve.py]: why `\r\n` and no unix newlines?

**Comment by EnTeQuAk** [comment]
Looks good to me, nice code improvements along the way.

File: src/olympia/editors/management/commands/auto_approve.py (hunk 2 above)
Comment (diox) [inline src/olympia/editors/management/commands/auto_approve.py]: I stole this from one of the `CannedResponse`s, and it had `\r\n`, so I kept them. (I couldn't easily fetch and use the `CannedResponse` programmatically as there is no way to identify it)

File: src/olympia/editors/management/commands/auto_approve.py (hunk 1 above)
Comment (diox) [inline src/olympia/editors/management/commands/auto_approve.py]: I debated that, but I preferred it didn't pollute the main rules.

Comment (EnTeQuAk) [inline src/olympia/editors/management/commands/auto_approve.py]: fair enough, I don't have any good reasons too, just wondered :)

**Pull request merged**